presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CrossJoinWithArrayContainsToInnerJoin.java

```diff
@@ -0,0 +1,181 @@
+        if (!(joinNode.getType().equals(JoinNode.Type.INNER) && joinNode.getCriteria().isEmpty())) {
+            return Result.empty();
+        }
+        List<VariableReferenceExpression> leftInput = joinNode.getLeft().getOutputVariables();
+        List<VariableReferenceExpression> rightInput = joinNode.getRight().getOutputVariables();
+        RowExpression filterExpression = node.getPredicate();
+        RowExpression arrayContainsExpression = getCandidateArrayContainsExpression(filterExpression, leftInput, rightInput);
+        if (arrayContainsExpression == null) {
+            return Result.empty();
+        }
+        List<RowExpression> andConjuncts = extractConjuncts(filterExpression);
+        List<RowExpression> remainingConjuncts = andConjuncts.stream().filter(x -> !x.equals(arrayContainsExpression)).collect(toImmutableList());
+
+        VariableReferenceExpression array = (VariableReferenceExpression) ((CallExpression) arrayContainsExpression).getArguments().get(0);
+        VariableReferenceExpression element = (VariableReferenceExpression) ((CallExpression) arrayContainsExpression).getArguments().get(1);
+        boolean arrayAtLeftInput = leftInput.contains(array);
+        PlanNode inputWithArray = arrayAtLeftInput ? joinNode.getLeft() : joinNode.getRight();
+
+        CallExpression arrayDistinct = call(functionAndTypeManager, "array_distinct", new ArrayType(element.getType()), array);
+        VariableReferenceExpression arrayDistinctVariable = context.getVariableAllocator().newVariable(arrayDistinct);
+        PlanNode project = PlannerUtils.addProjections(inputWithArray, context.getIdAllocator(), ImmutableMap.of(arrayDistinctVariable, arrayDistinct));
+        VariableReferenceExpression unnestVariable = context.getVariableAllocator().newVariable("field", element.getType());
+        UnnestNode unnest = new UnnestNode(inputWithArray.getSourceLocation(),
+                context.getIdAllocator().getNextId(),
+                project,
+                project.getOutputVariables(),
+                ImmutableMap.of(arrayDistinctVariable, ImmutableList.of(unnestVariable)),
+                Optional.empty());
+
+        JoinNode.EquiJoinClause equiJoinClause = arrayAtLeftInput ? new JoinNode.EquiJoinClause(unnestVariable, element) : new JoinNode.EquiJoinClause(element, unnestVariable);
+
+        JoinNode newJoinNode = new JoinNode(joinNode.getSourceLocation(),
+                context.getIdAllocator().getNextId(),
+                joinNode.getType(),
+                arrayAtLeftInput ? unnest : joinNode.getLeft(),
+                arrayAtLeftInput ? joinNode.getRight() : unnest,
+                ImmutableList.of(equiJoinClause),
+                joinNode.getOutputVariables(),
+                joinNode.getFilter(),
+                Optional.empty(),
+                Optional.empty(),
+                joinNode.getDistributionType(),
+                joinNode.getDynamicFilters());
+
+        if (!remainingConjuncts.isEmpty()) {
```

Comment:
Can we just add these as the filter to the JOIN? Should be OK as this is an INNERJOIN

Reply:
Good point. Just does this belong to a separate optimization rule?